gadgets/src/poseidon/mod.rs

```diff
@@ -0,0 +1,438 @@
+#![allow(non_snake_case)]
+
+use crate::poseidon::builder::Poseidon;
+use curve25519_dalek::scalar::Scalar;
+use bulletproofs::r1cs::{ConstraintSystem, R1CSError, Prover, Verifier};
+use bulletproofs::{PedersenGens};
+use bulletproofs::r1cs::LinearCombination;
+use crate::utils::{AllocatedScalar, constrain_lc_with_scalar};
+
+
+
+pub mod sbox;
+pub use sbox::*;
+pub mod builder;
+pub use builder::*;
+
+#[cfg(test)]
+pub mod tests;
+
+
+fn Poseidon_permutation(
+	input: &[Scalar],
+	params: &Poseidon,
+) -> Vec<Scalar>
+{
+	let width = params.width;
+	assert_eq!(input.len(), width);
+
+	let full_rounds_beginning = params.full_rounds_beginning;
+	let partial_rounds = params.partial_rounds;
+	let full_rounds_end = params.full_rounds_end;
+
+	let mut current_state = input.to_owned();
+	let mut current_state_temp = vec![Scalar::zero(); width];
+
+	let mut round_keys_offset = 0;
+
+	// full Sbox rounds
+	for _ in 0..full_rounds_beginning {
+		// Sbox layer
+		for i in 0..width {
+			current_state[i] += params.round_keys[round_keys_offset];
```

Comment:
Just as a note, I was doing this wrong: https://github.com/edgeware-builders/anon/blob/master/pallets/merkle/src/merkle/poseidon.rs#L141